test_cases/solidity/private_data_on_chain/odd_even_fixed/odd_even_fixed.sol

```diff
@@ -0,0 +1,86 @@
+/*
+ * @source: https://github.com/yahgwai/rps
+ * @author: Chris Buckland
+ * Modified by Kaden Zipfel
+ */
+
+pragma solidity ^0.5.0;
+
+contract OddEven {
+    enum Stage {
+        FirstCommit,
+        SecondCommit,
+        FirstReveal,
+        SecondReveal,
+        Distribution
+    }
+
+    struct Player {
+        address addr;
+        bytes32 commitment;
+        uint number;
+    }
+
+    Player[2] private players;
+    Stage public stage = Stage.FirstCommit;
+
+    function play(bytes32 commitment) public payable {
+        // Only run during commit stages
+        uint playerIndex;
+        if(stage == Stage.FirstCommit) playerIndex = 0;
+        else if(stage == Stage.SecondCommit) playerIndex = 1;
+        else revert("only two players allowed");
+
+        // Require proper amount deposited
+        // 1 ETH as a bet + 1 ETH as a bond
+        require(msg.value == 2 ether, 'msg.value must be 2 eth');
+
+        // Store the commitment
+        players[playerIndex] = Player(msg.sender, commitment, 0);
+
+        // Move to next stage
+        if(stage == Stage.FirstCommit) stage = Stage.SecondCommit;
+        else stage = Stage.FirstReveal;
+    }
+
+    function reveal(uint number, bytes32 blindingFactor) public {
+        // Only run during reveal stages
+        require(stage == Stage.FirstReveal || stage == Stage.SecondReveal, "wrong stage");
+
+        // Find the player index
+        uint playerIndex;
+        if(players[0].addr == msg.sender) playerIndex = 0;
+        else if(players[1].addr == msg.sender) playerIndex = 1;
+        else revert("unknown player");
+
+        // Check the hash to prove the player's honesty
+        require(keccak256(abi.encodePacked(msg.sender, number, blindingFactor)) == players[playerIndex].commitment, "invalid hash");
+
+        // Update player number if correct
+        players[playerIndex].number = number;
+
+        // Move to next stage
+        if(stage == Stage.FirstReveal) stage = Stage.SecondReveal;
+        else stage = Stage.Distribution;
+    }
+
+    function distribute() public {
+        // Only run during distribution stage
+        require(stage == Stage.Distribution, "wrong stage");
+
+        // Find winner
+        uint n = players[0].number + players[1].number;
+
+        // Payout winners winnings and bond
+        (bool success, ) = players[n%2].addr.call.value(3 ether)("");
+        require(success, "transfer failed");
```

Comment:
I think we should not require this to succeed because if it fails then we are stuck in this state.

Reply:
But if it fails, then the state is reset and the ETH is locked in the contract, right? What would be the alternative?

Reply:
Wouldn't it be better to not require it to succeed and ensure that the distribute() function completes regardsless if the external calls fail?

Reply:
If the calls fail, won't the users not receive the winnings and bonds? I think there's something I'm not understanding here...

Reply:
Let's say player 1 wins and player 2 loses and we are in Stage.Distribution. Player 2 could perform a griefing attack and let the call to transfer the bond back always fail. His bond would be stuck but so would be the winnings and the bond for player 1.

Reply:
That makes sense, thanks for explaining. Can you think of a way that we can assure the players are paid out and there's no griefing attack? I'll see if I can think of something tomorrow.

Reply:
If the players use EOA then the call always succeeds but with a contract it might not be so. Hmm I can't really think of a simple way to ensure that that the bond is transferred if the contract is deployed with a fallback function that fails.

Reply:
Only thing I can think of is only allowing EOAs to play. Of course this limits functionality, but I imagine this is the best solution? It's also starting to really complicate what's supposed to be a simple example, so I wonder if it might be worth coming up with a simpler contract as an example. What do you think?

Reply:
I think the only drawback from **not** requiring the call to succeed is that some contract that throws in its fallback function or that does not allow Ether to be sent to it might lose its bond. IMO this is acceptable. I think leaving a possibility where the contract can be stuck in a state is generally undesirable. 

So my proposition is to remove the require statements that check the call status, recompile the contracts and get it merged. Up for discussion.